src/Microsoft.AspNetCore.OData/Query/IQueryableODataExtension.cs

```diff
@@ -0,0 +1,218 @@
+        {
+            if (source is null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            foreach (var item in source)
+            {
+                yield return OCast<TResult>(item, options);
+            }
+        }
+
+        /// <summary>
+        /// Create the object based on <see cref="SelectExpandWrapper"/>
+        /// </summary>
+        /// <param name="wrapper">The select expand wrapper.</param>
+        /// <param name="resultType">The created type.</param>
+        /// <param name="options">The options.</param>
+        /// <returns>The created object.</returns>
+        private static object CreateInstance(SelectExpandWrapper wrapper, Type resultType, ODataCastOptions options)
+        {
+            object instance;
+            Type instanceType = resultType;
+            IEdmModel model = wrapper.Model;
+            if (wrapper.UseInstanceForProperties)
+            {
+                instance = wrapper.InstanceValue;
+                instanceType = instance.GetType();
+            }
+            else
+            {
+                if (wrapper.InstanceType != null && wrapper.InstanceType != resultType.FullName)
+                {
+                    // inheritance
+                    IEdmTypeReference typeReference = wrapper.GetEdmType();
+                    instanceType = model.GetClrType(typeReference); // inheritance type
+                }
+
+                instance = Activator.CreateInstance(instanceType);
+            }
+
+            IDictionary<string, object> properties;
+            if (options != null && options.MapProvider != null)
+            {
+                properties = wrapper.ToDictionary(options.MapProvider);
+            }
+            else
+            {
+                properties = wrapper.ToDictionary();
+            }
+
+            foreach (var property in properties)
+            {
+                string propertyName = property.Key;
+                object propertyValue = property.Value;
+                if (propertyValue == null)
+                {
+                    // If it's null, we don't need to do anything.
+                    continue;
+                }
+
+                PropertyInfo propertyInfo = GetPropertyInfo(instanceType, propertyName);
+                if (propertyInfo == null)
+                {
+                    throw new ODataException(Error.Format(SRResources.PropertyNotFound, instanceType.FullName, propertyName));
+                }
+
+                bool isCollection = TypeHelper.IsCollection(propertyInfo.PropertyType, out Type elementType);
+
+                if (isCollection)
+                {
+                    IList<object> collection = new List<object>();
+                    IEnumerable collectionPropertyValue = propertyValue as IEnumerable;
+                    foreach (var item in collectionPropertyValue)
```

Comment:
This is a mistake. `as` should not be used when the semantic is "the value _must_ be of the type", since it can return `null` and this will cause a `NullReferenceExeception` in the `foreach`.

```suggestion
                    IEnumerable collectionPropertyValue = (IEnumerable)propertyValue;
                    foreach (var item in collectionPropertyValue)
```